File: grading/grading.go
package grading

func gradingStudents(grades []int32) []int32 {
	var result []int32

	for _, grade := range grades {
		if grade < 38 {
			result = append(result, grade)
			continue
		}

		multiple := nextMultiple(5, grade)
		if multiple-grade >= 3 {
			result = append(result, grade)
			continue
		}

		result = append(result, multiple)
	}

	return result
}

func nextMultiple(denom int32, n int32) int32 {
	next := n + 1

	if next%denom == 0 {
		return next
	} else {
		return nextMultiple(denom, next)
	}
}

File: grading/Grading.java
import java.util.List;
import java.util.stream.Collectors;

public class Grading {

  public static List<Integer> gradingStudents(List<Integer> grades) {
    return grades
      .stream()
      .map(grade -> {
        if (grade < 38) {
          return grade;
        }

        int multiple = nextMultiple(5, grade);
        if (multiple - grade >= 3) {
          return grade;
        }

        return multiple;
      })
      .collect(Collectors.toList());
  }

  private static int nextMultiple(int denom, int n) {
    int next = n + 1;

    return next % denom == 0
      ? next
      : nextMultiple(denom, next);
  }

}